### PR TITLE
Change slightly how `x`/`X` works

### PR DESCRIPTION
**Pull request by dpc**
Discussion in #2590

### CODE REVIEW

File: src/context.hh [hunk 1]
```diff
@@ -135,6 +148,8 @@ private:
     void end_edition();
     int m_edition_level = 0;
     size_t m_edition_timestamp = 0;
+    mutable bool m_line_mode = false;
+    mutable bool m_keep_line_mode = false;
```
[the remark this answers is not in the document]
Comment (dpc) [reply src/context.hh]: It's kind of a hack/workaround that `mawww` throw at me, when I didn't know what to do. You can revert `Is it a hack, or is it a solution? :D`, or see the error in this commit: https://github.com/dpc/kakoune/commit/b8d3f08a20f7470d894cd562d1ef24bebff9ac76

**Comment by maximbaz** [comment]
#2590 became a broad discussion about different ideas, let's make it super clear that this PR is about one thing: changing when `x/X` move to the next line, from "if whole line is already selected" to "if the previous action was also `x/X`". @dpc maybe add this to the first message of the PR to clarify.

@mawww would be interesting to hear your opinion about this particular proposal.

File: src/normal.cc [hunk 2]
```diff
@@ -2001,7 +2001,12 @@ class Repeated
     void operator() (Context& context, NormalParams params)
     {
         ScopedEdition edition(context);
-        do { m_func(context, {0, params.reg}); } while(--params.count > 0);
+        do {
+            m_func(context, {0, params.reg});
+            if (params.count > 1) {
+                context.post_movement_logic();
+            }
```
[the remark this answers is not in the document]
Comment (dpc) [reply src/normal.cc]: Gosh. I think it's a terrible idea, but not my project, not my rules. Fixing. :)

**Comment by dpc** [comment]
I have fixed the style. I was editing this with kakoune, that I haven't configured for cpp yet, so there were some troubles and omissions. :)

File: src/normal.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (lenormf) [reply src/normal.cc]: Would need to be fixed in subsequent hunks as well, prior to merging.

**Comment by andreyorst** [comment]
About several formatting suggestions mentioned in reviews. Just wondering, why not set up git hooks or add file with rules for formatting programm (clang-format for example). It works for such projects as Linux, could work here. And kakoune supports formatting, so it feels natural to have project formatting rules.

**Comment by mawww** [comment]
I must say I am not really in favor of that change, my main issue is that this de-facto introduces a 'submode' into the normal mode, and the behaviour is pretty fuzzy, as it is context dependent.

Say if you do `x` in client one, then the selected line is joined to the next in client two, so the selection is not a full line anymore in client one, and then you press x again in client one, you get to the next line without having ever fully selected the current one.

In general, I think introducing this additional state is confusing and creates a dangerous precedent. I would love to improve the ergonomics of `x`, but I am not confortable with this approach.

**Comment by dpc** [comment]
@andreyorst Yeah, I was wondering the same thing. Rust spoiled me.

@mawww I understand. In practice I don't hit this fuzziness with this branch, while I get supper annoyed by having inconsistent behavior on empty lines.

In this one particular case you mention, should this state move to a different data structure that is shared between clients? Where it would be?

**Comment by mawww** [comment]
> In this one particular case you mention, should this state move to a different data structure that is shared between clients? Where it would be?

Would'nt that be even worse ? If you press `x` in one client, get to another client and press `x`, you magically jump to the next line without having ever selected the current one (and clients might be at totally different positions in the buffer, or on different buffers), now imagine what happens with scripting or just mappings: `x` becomes even less predictible as it depends on if the user has hit `x` before triggering the mapping or not.

**Comment by dpc** [comment]
I don't know the details of buffer sharing between clients, so I have no idea which one is better.

The state of `x` consecutive-ness could be tracked per client, per script connection etc. too.

**Pull request closed without merging**